base_layer/wallet/src/transaction_service/storage/models.rs

```diff
@@ -237,6 +244,117 @@ impl CompletedTransaction {
+
+        // Calculate average features and scripts size from actual transaction outputs
+        let original_outputs = self.transaction.body.outputs();
+        let total_features_and_scripts_size: Result<usize, std::io::Error> = original_outputs
+            .iter()
+            .map(|output| output.get_features_and_scripts_size())
+            .sum();
+
+        let total_features_and_scripts_size = total_features_and_scripts_size.map_err(|e| {
+            TransactionStorageError::FailedToCalculateTransactionFee(format!(
+                "Failed to calculate features and scripts size from original transaction: {}",
+                e
+            ))
+        })?;
+
+        // Calculate average size per output from original transaction
+        let avg_features_and_scripts_size_per_output = if original_outputs.is_empty() {
+            // Fallback to default if no outputs (shouldn't happen, but just in case)
+            let default_output_features_size = OutputFeatures::default().get_serialized_size().map_err(|e| {
+                TransactionStorageError::FailedToCalculateTransactionFee(format!(
+                    "Failed to serialize OutputFeatures: {}",
+                    e
+                ))
+            })?;
+            let default_script_size = TariScript::default().get_serialized_size().map_err(|e| {
+                TransactionStorageError::FailedToCalculateTransactionFee(format!(
+                    "Failed to serialize TariScript: {}",
+                    e
+                ))
+            })?;
+            let default_covenant_size = Covenant::new().get_serialized_size().map_err(|e| {
+                TransactionStorageError::FailedToCalculateTransactionFee(format!("Failed to serialize Covenant: {}", e))
+            })?;
+            default_output_features_size + default_script_size + default_covenant_size
+        } else {
+            total_features_and_scripts_size / original_outputs.len()
+        };
+
+        // Apply rounding and multiply by number of outputs for new transaction
+        let features_and_scripts_size = fee_calculator
+            .weighting()
+            .round_up_features_and_scripts_size(avg_features_and_scripts_size_per_output) *
+            num_outputs;
+
+        // Use the Fee struct's weighting calculation to get transaction weight in grams
+        let weight_in_grams = fee_calculator.weighting().calculate(
+            1, // num_kernels
+            num_inputs,
+            num_outputs,
+            features_and_scripts_size,
+        );
+
+        let fee_per_gram = if weight_in_grams > 0 {
+            let fee_per_gram_f64 = total_fee.0 as f64 / weight_in_grams as f64;
+            #[allow(clippy::cast_possible_truncation)]
+            let fee_per_gram_u64 = fee_per_gram_f64.round() as u64;
+            MicroMinotari::from(fee_per_gram_u64)
+        } else {
+            MicroMinotari::zero()
+        };
+
+        Ok((weight_in_grams, fee_per_gram))
```

Comment:
Consider adding a sanity check to ensure the new fee request is met, for example.
```suggestion
        let mut fee_per_gram = MicroMinotari::from(fee_per_gram);
        loop {
            let calculated_fee = fee_calculator.calculate(
                fee_per_gram,
                1, // num_kernels = 1
                num_inputs,
                num_outputs,
                features_and_scripts_size,
            );

            if calculated_fee >= total_fee {
                break;
            }

            fee_per_gram += MicroMinotari::from(1);
        }

        Ok((weight_in_grams, fee_per_gram))
```

Reply:
I am bit reluctant to implement this - if calculation where indeed invalid than we would get an error and be aware that there is something off with a fee calculations. But with this sanity check we would cover up internal calculation bugs.

Reply:
Added loop as final validation of whole calculation to ensure the calculated fee is higher or equal to that given as function argument.

Reply:
I think remove the loop, its unnessary complex and will cause a bit of loops. 
Even if we have the perfect fee per gram here, we are not going to end up with the perfect fee on the other end, keep this best effort